sway-core/src/transform/to_parsed_lang/convert_parse_tree.rs

```diff
@@ -3933,3 +3988,54 @@ fn error_if_self_param_is_not_allowed(
     }
     Ok(())
 }
+
+/// Walks all the cfg attributes in a map, evaluating them
+/// and returning false if any evaluated to false.
+pub fn cfg_eval(
+    context: &mut Context,
+    handler: &Handler,
+    attrs_map: &AttributesMap,
+) -> Result<bool, ErrorEmitted> {
+    if let Some(cfg_attrs) = attrs_map.get(&AttributeKind::Cfg) {
+        for cfg_attr in cfg_attrs {
+            for arg in &cfg_attr.args {
+                match arg.name.as_str() {
+                    "target" => {
+                        if let Some(value) = &arg.value {
+                            if let sway_ast::Literal::String(value_str) = value {
+                                if let Ok(target) = BuildTarget::from_str(value_str.parsed.as_str())
+                                {
+                                    if target != context.build_target() {
+                                        return Ok(false);
+                                    }
+                                } else {
+                                    let error = ConvertParseTreeError::InvalidCfgTargetArgValue {
+                                        span: value.span(),
+                                        value: value.span().str(),
+                                    };
+                                    return Err(handler.emit_err(error.into()));
+                                }
+                            } else {
+                                let error = ConvertParseTreeError::InvalidCfgTargetArgValue {
+                                    span: value.span(),
+                                    value: value.span().str(),
+                                };
+                                return Err(handler.emit_err(error.into()));
+                            }
+                        } else {
+                            let error = ConvertParseTreeError::ExpectedCfgTargetArgValue {
+                                span: arg.span(),
+                            };
+                            return Err(handler.emit_err(error.into()));
+                        }
+                    }
+                    _ => {
+                        // Already checked with `AttributeKind::expected_args_*`
+                        unreachable!("cfg attribute should only have the target argument");
```

Comment:
Would certainly be nice to have negative tests here to make sure we error out nicely.